harness/tests/determined/pytorch/experimental/test_torch_batch_process.py

```diff
@@ -0,0 +1,543 @@
+        self.context = context
+        self.reducer_sum = context.wrap_reducer(reducer=MySumMetricReducer(), name="sum_metric")
+        self.rank = context.distributed.get_rank()
+
+    def process_batch(self, batch: Any, batch_idx: int) -> None:
+        self.reducer_sum.update(batch)
+        print(f"My rank is {self.rank} and current batch is {batch_idx}")
+
+
+class IndexData(data.Dataset):
+    def __init__(self, data_length: int = 50) -> None:
+        self.data = data_length
+
+    def __len__(self) -> int:
+        return int(self.data)
+
+    def __getitem__(self, idx: int) -> int:
+        return idx
+
+
+def _get_core_context(
+    rank: int = 0, should_preempt_results: Optional[List[bool]] = None
+) -> unittest.mock.MagicMock:
+    mock_core_context = unittest.mock.MagicMock()
+    mock_distributed_context = _get_dist_context(rank=rank)
+    mock_core_context.__enter__().distributed = mock_distributed_context
+    if should_preempt_results is None:
+        mock_core_context.__enter__().preempt.should_preempt.return_value = False
+    else:
+        mock_core_context.__enter__().preempt.should_preempt.side_effect = should_preempt_results
+    return mock_core_context
+
+
+def _get_dist_context(
+    rank: int = 0,
+    all_gather_return_value: Optional[Any] = None,
+    gather_return_value: Optional[Any] = None,
+) -> unittest.mock.MagicMock:
+    mock_distributed_context = unittest.mock.MagicMock()
+    mock_distributed_context.get_rank.return_value = rank
+    mock_distributed_context.broadcast.return_value = "mock_checkpoint_uuid"
+    mock_distributed_context.allgather.return_value = all_gather_return_value
+    mock_distributed_context.gather.return_value = gather_return_value
+    return mock_distributed_context
+
+
+@pytest.mark.parametrize(
+    "dataloader_kwargs, batch_size",
+    [
+        [{"shuffle": True}, 10],
+        [{"sampler": unittest.mock.Mock()}, 10],
+        [{"batch_sampler": unittest.mock.Mock()}, 10],
+        [{"batch_size": 20}, 10],
+    ],
+)
+def test_torch_batch_process_dataloader_kwargs_validation(
+    dataloader_kwargs: Dict[str, Any],
+    batch_size: int,
+) -> None:
+    # Test calling torch_batch_process with invalid arguments
+    with pytest.raises(ValueError):
+        experimental._torch_batch_process._validate_dataloader_kwargs(dataloader_kwargs, batch_size)
+
+
+@unittest.mock.patch(
+    "determined.pytorch.experimental._torch_batch_process._initialize_default_inference_context"
+)
+@unittest.mock.patch(
+    "determined.pytorch.experimental._torch_batch_process._synchronize_and_checkpoint"
+)
+@unittest.mock.patch(
+    "determined.pytorch.experimental._torch_batch_process._report_progress_to_master"
+)
+@pytest.mark.parametrize(
+    "data_length, batch_size, checkpoint_interval,rank, num_slots",
+    [  # Group 1: data length is divisible by batch_size
+        [50, 10, 1, 0, 2],  # multi-slots, chief
+        [50, 10, 1, 1, 2],  # multi-slots, non-chief
```

Comment:
@rb-determined-ai I cannot use the parallel unit in the test_util as patch function of unittest is not thread safe, I saw non-deterministic call count when I run tests of chief and non-chief worker in parallel. 

The other cases in our code base that use mock with parallel run are not using patch mechanism.

Reply:
I mean, sure. But also your assertion isn't necessary if you use parallel.Execution, because parallel.Execution would hang if the _synchronize_and_checkpoint were called the wrong number of times on some worker.

Reply:
That is a good point, I will keep it as is for now as they test for the same thing and failure reason may be clearer if the call count differs as opposed to just hanging.